File: Sort-Array-By-Parity-II/SABPII.rs
pub fn sort_array_by_parity_ii(a: Vec<i32>) -> Vec<i32> {
    let mut aa = a.clone();
    aa.sort();
    let mut even = aa.iter().filter(|&x| *x % 2 == 0);
    let mut odd = aa.iter().filter(|&x| *x % 2 != 0);

    let mut result = vec![];
    for _ in 0..a.len() / 2 {
        result.push(*even.next().unwrap());
        result.push(*odd.next().unwrap());
    }
    result
}

fn main() {
    dbg!(sort_array_by_parity_ii(vec![4, 2, 5, 7]));
}

File: Sort-Array-By-Parity-II/SABPII.hs
import           Data.List (sort)

sort_array_by_parity_ii :: [Int] -> [Int]
sort_array_by_parity_ii ll =
  let even_l = filter even ll
      odd_l = filter odd ll
  in
    merge_ll (sort even_l) (sort odd_l)
  where
    merge_ll [] y          = y
    merge_ll x []          = x
    merge_ll (x:xs) (y:ys) = x:y:(merge_ll xs ys)

main = do
  print $ sort_array_by_parity_ii [4,2,5,7]
